docs/dev_plan.md

```diff
@@ -0,0 +1,339 @@
+The key concept of this paradigm is to structure data so that it maps
+closely to the form it will be represented and used within the relevant
+processing units. For example, it is a common mistake in object-oriented
+design to represent data as arrays of structures (AoS),
+and this can lead to inefficiency in CPU cache
+usage since aggregate operations such as matrix math must load
+data from non-contiguous memory rather than perform the operations
+in a vector form. In a data-oriented design, data is more
+commonly represented as structures of arrays (SoA) where contiguous
+portions of memory are loaded into the cache and operated on
+either by the CPU or a hardware accelerator. The graphic below
+illustrates the difference between structures of arrays and arrays
+of structures.
+
+```{image} images/AoS_SoA.pdf
+:alt: aos_soa
+:width: 400px
+:align: center
+```
+
+The difference in usage for the above two patterns is illustrated via
+an example Python-based math operation below where a series of points
+are shifted in one direction by a constant amount.
+
+```python
+class Point:
+    x: int
+    y: int
+    z: int
+
+# Initialize test data
+all_points = [Point(random(i), random(i), random(i)) for i in range(100)]
+
+# Array of Structures
+AOS = [all_points[i].y for i in range(100)]
+
+# Structure of Arrays
+SOA = all_points.y + 10
+```
+
+While the AoS vs SoA pattern is a typical example of data-oriented vs object-oriented modeling,
+the concept extends to all areas of data modeling in software and is equally relevant to most
+programming languages.
+
+### Accessible software
+
+Access to the software is an important consideration for the
+longevity and relevance of OpenTurbine. In short, if the software is
+not accessible, it won't be used, extended, or maintained. All
+development efforts should always consider user and developer accessibility
+as a key driver. The distinction between "user" and "developer"
+is not always clear, but accessibility efforts should address concerns
+for both types of engagement with OpenTurbine.
+
+Documentation is the primary tool for addressing accessibility. As a
+guideline, the practice of documentation driven development should be
+widely adopted here. For any new feature or change, describe it first
+in the documentation. New features should be outlined in text to describe
+the high level objectives. Algorithmic development and design decisions
+should be included directly in the developer documentation. Before any
+code is written, the workflows in which the new code will be used should
+be expressed resulting in a proposed API. Finally, unit tests should be
+written to match the feature description in the documentation. Upon 
+implementation and subsequent iterations, the unit tests should retain
+their close correlation to the documentation so that any deviation
+between the implementation and documentation is caught through the tests.
```

Comment:
This is exactly the problem, unit-tests don't catch any deviation from documentation. How do you unit-test documentation? Unit-tests necessarily track the code. When any deviation happens, the code is the truth and the documentation is a lie.